### PR TITLE
Fix manipulation of PathPoint(s) invoked from OpenSimContext for Path editing

### PR DESCRIPTION
**Pull request by aymanhab**
So far deletion only is tested, rest will come later in this PR but investigation into this can start now and will clarify what to change in 4.0

Fixes issue #1670

### Brief summary of changes
In the past, changes to GeometryPath didn't require recreating the system under the model but rather to re-realize the state. Now that PathPoints are Components and are part of the system, edits by insertion, deletion or typechange require recreating the system.

### Testing I've completed
testContext has been made stronger to make the calls done by GUI path editing. This guarantees the calls don't crash or throw anymore.
Testing addition, removal and/or type change from GUI works with this PR in place, though MovingPathPoints have not been thoroughly  tested.

### Looking for feedback on...
The change of type is adhoc and doesn't cover all the combination of pairs of types of PathPoints, Need more exercise from GUI.
 
### CHANGELOG.md (choose one)

- no need to update because this is internal bug-fix. 

The Doxygen for this PR can be viewed at http://myosin.sourceforge.net/?C=N;O=D; click the folder whose name is this PR's number.

### CODE REVIEW

File: Bindings/Java/OpenSimJNI/OpenSimContext.cpp [hunk 1]
```diff
@@ -264,7 +264,7 @@ void OpenSimContext::addPathPoint(GeometryPath& p, int menuChoice, PhysicalFrame
 
 bool OpenSimContext::deletePathPoint(GeometryPath& p, int menuChoice) {
     bool ret = p.deletePathPoint(*_configState, menuChoice );
-  _configState->invalidateAll(SimTK::Stage::Position);
+  _configState->invalidateAll(SimTK::Stage::Position); // This maynot be enough anymore in 4.0
```
Comment (aseth1) [inline Bindings/Java/OpenSimJNI/OpenSimContext.cpp]: Deleting, adding, replacing, etc the `PathPoint` changes the model. The System should no longer be valid, in general. Look at your code for `replacePathPoint` that recreates the System after the edit.

**Comment by aymanhab** [comment]
Ready for review.

File: Bindings/Java/OpenSimJNI/OpenSimContext.cpp [hunk 2]
```diff
@@ -243,30 +233,26 @@ void OpenSimContext::setLocation(PathPoint& mp, int i, double d) {
...
+    p.updateGeometry(*_configState);
     return;
 }
 
 bool OpenSimContext::deletePathPoint(GeometryPath& p, int menuChoice) {
     bool ret = p.deletePathPoint(*_configState, menuChoice );
-  _configState->invalidateAll(SimTK::Stage::Position);
-  _model->getMultibodySystem().realize(*_configState, SimTK::Stage::Position);
-  return ret;
+    recreateSystemKeepStage();
```
Comment (tkuchida) [inline Bindings/Java/OpenSimJNI/OpenSimContext.cpp]: Not sure I understand the logic here: why is `recreateSystemKeepStage()` called even if the PathPoint can't be deleted? Wouldn't the following be more efficient?
``` cpp
if (!p.canDeletePathPoint(menuChoice))
    return false;

const bool ret = p.deletePathPoint(*_configState, menuChoice);
recreateSystemKeepStage();
return ret;
```

Comment (aymanhab) [inline Bindings/Java/OpenSimJNI/OpenSimContext.cpp]: Nice catch @tkuchida, it was more of blanket code transformation to replace the invalidate/realize call pair with recreateSystemKeepStage but indeed that is not necessary in this case :+1: Will make the change, thank you.

**Comment by aymanhab** [comment]
Build failure is an unrelated  timeout. Thanks @tkuchida Any feedback @aseth1 ? Thanks

**Comment by aymanhab** [comment]
@aseth1 any feedback to report, this is pretty straightforward code transformation to reflect that any changes to wiring or properties require recreation of system rather than invalidating stage, or if you can take a look @chrisdembia that would be great. Thanks.

File: Bindings/Java/OpenSimJNI/Test/testContext.cpp [hunk 3]
```diff
@@ -187,6 +187,41 @@ int main()
...
+    assert(pathPoints.getSize() == origSize);
+    std::string pathAfterReinsertionInXML = dTRIlong->updGeometryPath().dump();
+    std::cout << pathAfterReinsertionInXML << endl;
+
+    // Test changing type to ConditionalPathPoint
+    ConditionalPathPoint* newPoint = new ConditionalPathPoint();
+    AbstractPathPoint& oldPoint = pathPoints.get(2);
+    newPoint->setCoordinate(model->getCoordinateSet().get(0));
+    newPoint->setParentFrame(oldPoint.getParentFrame());
+    context->replacePathPoint(dTRIlong->updGeometryPath(), oldPoint, *newPoint);
```
Comment (aseth1) [inline Bindings/Java/OpenSimJNI/Test/testContext.cpp]: The requirement of `replacePathPoint()` takes a reference to the oldPoint is a bit troubling. This pretty much guarantees you have a stale reference lying around. Your `delete` method takes an index, and you get the point by index above, so would it not be safer here to also take the index to be replaced?

**Comment by aseth1** [comment]
The changes LGTM to me. I only have one issue above, but it was not created by this PR. It could be addressed secondarily if not convenient to change herein.

File: Bindings/Java/OpenSimJNI/Test/testContext.cpp (hunk 3 above)
Comment (aymanhab) [inline Bindings/Java/OpenSimJNI/Test/testContext.cpp]: Thanks for the review @aseth1 
Good point regarding references (actually most OpenSim objects live as references on scripting & GUI side) I'm trying to keep this PR though limited to change in the implementation without change to the API.  I'll open one general one regarding the use of references.

**Comment by aymanhab** [comment]
issue #1701 opened to address feedback on use of references.

**Pull request merged**